strawberryfields/io.py

```diff
@@ -0,0 +1,214 @@
+    """
+    # create a SF program
+    if not bb.modes:
+        # we can't return an empty program, since we don't know how many modes
+        # to initialize the Program object with.
+        raise ValueError("Blackbird program contains no quantum operations!")
+
+    prog = Program(max(bb.modes)+1, name=bb.name)
+
+    # append the quantum operations
+    with prog.context as q:
+        for op in bb.operations:
+            try:
+                if 'args' in op:
+                    getattr(ops, op["op"])(*op["args"], **op["kwargs"]) | [q[i] for i in op["modes"]] #pylint:disable=expression-not-assigned
+                else:
+                    getattr(ops, op["op"]) | [q[i] for i in op["modes"]] #pylint:disable=expression-not-assigned
+            except AttributeError:
+                raise NameError("Quantum operation {} not defined!".format(op["op"]))
+
+    # compile the program if a target exists
+    if bb.target["name"] is not None:
+        return prog.compile(bb.target["name"])
+
+    return prog
+
+
+def save(file, prog):
+    """Saves a quantum program to a Blackbird .xbb file.
+
+    Args:
+        file (Union[file, str, pathlib.Path]): File or filename to which
+            the data is saved. If file is a file-object, then the filename
+            is unchanged. If file is a string or Path, a .xbb extension will
+            be appended to the file name if it does not already have one.
+        prog (Program): Strawberry Fields program
+    """
+    own_file = False
+    bb = to_blackbird(prog).serialize()
+
+    if hasattr(file, "read"):
+        # argument file is a file-object
+        fid = file
+    else:
+        # argument file is a string or Path
+        f = os.fspath(file)
+
+        if not f.endswith(".xbb"):
+            f = f + ".xbb"
+
+        fid = open(f, "w")
+
+        # this function owns the open file,
+        # must remember to safely close it.
+        own_file = True
+
+    try:
+        fid.write(bb)
+    finally:
+        if own_file:
+            # safely close the file
+            fid.close()
+
+
+def load(file):
+    """Load a quantum program from a Blackbird .xbb file.
+
+    Args:
+        file (Union[file, str, pathlib.Path]): File or filename to which
+            the data is saved. If file is a file-object, then the filename
+            is unchanged. If file is a string or Path, a .xbb extension will
+            be appended to the file name if it does not already have one.
+
+    Returns:
+        prog (Program): Strawberry Fields program
+    """
+    own_file = False
+
+    try:
+        if hasattr(file, "read"):
+            # argument file is a file-object
+            fid = file
+        else:
+            # argument file is a Path or string
+            f = os.fspath(file)
+            fid = open(f, "r")
+            own_file = True
+
+    except TypeError:
+        raise ValueError("file must be a string, pathlib.Path, or file-like object")
```

Comment:
Something feels wrong about this exception, specifically to do with the appearance of both `file` and `file-like`. I'm not really sure how to improve though.

Reply:
I copied this from the NumPy documentation, where they use the term `file-like` to describe any object that supports the `.write()` method.

Reply:
I'll change the argument to `f`?

Reply:
Done.